### PR TITLE
batik: refactor, modernize

### LINKED ISSUE #50932: Batik package should expose its jars as executables
The current batik package: https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/graphics/batik/default.nix

Exposes several executable jars:

* `batik.jar`
* `batik-rasterizer.jar`
* `batik-slideshow.jar`
* `batik-squiggle.jar`
* `batik-svgpp.jar`
* `batik-ttf2svg.jar`

All of these can be exposed CLI executables directly similar to the plantuml package.

### PR DESCRIPTION
**Pull request by tomodachi94**
## Description of changes

* Moved to `pkgs/by-name` and formatted with `nixfmt-rfc-style`
* Refactor the derivation:
  * All JARs end up in $out/share/java, and a few wrappers for the applications are made in $out/bin (closes #50932).
  * stdenv -> stdenvNoCC, since no compilers are needed in any stages.
  * Remove the vendored Rhino JavaScript engine and replace it with our Nixpkgs one.
  * Add the stripJavaArchivesHook.

### Considerations

Currently, `batik` and `fop` both vendor the `xmlgraphics-common` JAR. This should be refactored at a later date.

Our version of Rhino is old. _Really_ old. I wouldn't be surprised if it causes issues in this package; it should be updated at some point.

We could probably compile this package from source once we have all of the vendored dependencies packaged.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/ba/batik/package.nix [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  lib,
+  stdenvNoCC,
+  fetchurl,
+  jre,
+  rhino,
+  stripJavaArchivesHook,
+  makeWrapper,
+}:
+
+stdenvNoCC.mkDerivation rec {
```
Comment (afh) [inline pkgs/by-name/ba/batik/package.nix:11]: Use of `finalAttrs` is preferred over `rec` if this suggestion is accepted remember to close the newly opened parenthesis (see suggestion at the end of the pacckage)
```suggestion
stdenvNoCC.mkDerivation (finalAttrs: {
```

File: pkgs/by-name/ba/batik/package.nix [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  '';
+
+  meta = with lib; {
+    description = "Java based toolkit for handling SVG";
+    homepage = "https://xmlgraphics.apache.org/batik";
+    license = licenses.asl20;
+    platforms = platforms.unix;
+    sourceProvenance = with sourceTypes; [ binaryBytecode ];
+  };
+}
```
Comment (afh) [inline pkgs/by-name/ba/batik/package.nix:54]: Close the parenthesis opened for `finalAttrs` (see `stdenvNoCC.mkDerivation` line above):
```suggestion
})
```

File: pkgs/by-name/ba/batik/package.nix [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  makeWrapper,
+}:
+
+stdenvNoCC.mkDerivation rec {
+  pname = "batik";
+  version = "1.17";
+
+  src = fetchurl {
+    url = "mirror://apache/xmlgraphics/batik/binaries/batik-bin-${version}.tar.gz";
+    sha256 = "sha256-sEJphF3grlwZCEt3gHHm4JF8RpvKKBLLvKXf2lu/dhA=";
```
Comment (afh) [inline pkgs/by-name/ba/batik/package.nix:17]: Use of `hash` is preferred over `sha256` as `hash` provides more flexibility to use a different type of hash in the future, e.g. `sha512`, as nix hashes can include the hash-type has a prefix:
```suggestion
    hash = "sha256-sEJphF3grlwZCEt3gHHm4JF8RpvKKBLLvKXf2lu/dhA=";
```

File: pkgs/by-name/ba/batik/package.nix [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  stripJavaArchivesHook,
+  makeWrapper,
+}:
+
+stdenvNoCC.mkDerivation rec {
+  pname = "batik";
+  version = "1.17";
+
+  src = fetchurl {
+    url = "mirror://apache/xmlgraphics/batik/binaries/batik-bin-${version}.tar.gz";
```
Comment (afh) [inline pkgs/by-name/ba/batik/package.nix:16]:
```suggestion
    url = "mirror://apache/xmlgraphics/batik/binaries/batik-bin-${finalAttrs.version}.tar.gz";
```

File: pkgs/by-name/ba/batik/package.nix [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  '';
+
+  installPhase = ''
+    mkdir -p $out/bin $out/share/java
+    cp *.jar lib/*.jar $out/share/java
+    chmod +x $out/share/java/*.jar
+    classpath="$(find $out/share/java -name '*.jar' -printf '${rhino}/share/java/js.jar:%h/%f')"
+    for appName in rasterizer slideshow squiggle svgpp ttf2svg; do
+      makeWrapper ${jre}/bin/java $out/bin/batik-$appName \
+        --add-flags "-jar $out/share/java/batik-all-${version}.jar" \
```
Comment (afh) [inline pkgs/by-name/ba/batik/package.nix:41]:
```suggestion
        --add-flags "-jar $out/share/java/batik-all-${finalAttrs.version}.jar" \
```

File: pkgs/by-name/ba/batik/package.nix [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+  src = fetchurl {
+    url = "mirror://apache/xmlgraphics/batik/binaries/batik-bin-${version}.tar.gz";
+    sha256 = "sha256-sEJphF3grlwZCEt3gHHm4JF8RpvKKBLLvKXf2lu/dhA=";
+  };
+
+  nativeBuildInputs = [
+    stripJavaArchivesHook
+    makeWrapper
+  ];
```
Comment (afh) [inline pkgs/by-name/ba/batik/package.nix:23]: Most packages I've come across separate phases with an empty newline, would be nice to do it here too…
```suggestion
  ];

```

**Review by afh — CHANGES_REQUESTED** [review]
Thanks @tomodachi94 for refactoring and modernizing the batik package, please find below a few comments on suggestions for improvement.

Hopefully I've been able to explain well the motivation behind the suggestions, if not please ask me to clarify 🙂

**Comment by tomodachi94** [comment]
Good catches :sweat_smile: I'm embarrassed that I didn't catch all of that. Should be fixed in the latest push.

File: pkgs/by-name/ba/batik/package.nix [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    rm lib/rhino-*.jar
+  '';
+
+  installPhase = ''
+    mkdir -p $out/bin $out/share/java
+    cp *.jar lib/*.jar $out/share/java
+    chmod +x $out/share/java/*.jar
+    classpath="$(find $out/share/java -name '*.jar' -printf '${rhino}/share/java/js.jar:%h/%f')"
+    for appName in rasterizer slideshow squiggle svgpp ttf2svg; do
+      makeWrapper ${jre}/bin/java $out/bin/batik-$appName \
```
Comment (afh) [inline pkgs/by-name/ba/batik/package.nix:41]: Using `lib.getExe` would future-proof this package further:
```suggestion
      makeWrapper ${lib.getExe jre} $out/bin/batik-$appName \
```

**Review by afh — CHANGES_REQUESTED** [review]
Thanks for accepting the suggestions and making the changes, @tomodachi94, much appreciated 🙂 

Please find below another suggestion in case you wanted to go above and beyond.

Not catching all the things that can be catched happens to me all the time, hence this request for more changes 😅 (see comment below about `lib.getExe`) and I think that's just a prime example of why peer review is so beneficial 👏

File: pkgs/by-name/ba/batik/package.nix (hunk 7 above)
Comment (tomodachi94) [reply pkgs/by-name/ba/batik/package.nix]: Done, I love the `lib.getExe` pattern :)

**Review by afh — APPROVED** [review]
Thanks, @tomodachi94, for updating this PR, well done!

Approving as is, yet I'd appreciated of you could squash the 3 commits of this PR into a single one.

**Comment by tomodachi94** [comment]
> I'd appreciated of you could squash the 3 commits of this PR into a single one.

I opted not to squash the commits so that it's clear what happened to this file in the PR. Looking at the commits, you can see it was moved, formatted, then refactored. I've noticed that Git sometimes has trouble determining where a file went when it is moved and then edited in the same commit.

**Comment by afh** [comment]
My motivation to ask for a squash is to avoid git commit noise as the project has so many commits that reducing these has a certain benefit. That said I can also see the points you bring forward. I'll leave it to you to squash or leave as is.

**Comment by fpletz** [comment]
> Approving as is, yet I'd appreciated of you could squash the 3 commits of this PR into a single one.

IMHO those three commits strike the perfect balance when to squash and when not to. The by-name move and the reformat should be extra commits. Thanks! :+1:

**Comment by afh** [comment]
Thanks for chiming in on this, @fpletz, that's helpful 🙂

**Pull request merged**